src/escapei.rs

```diff
@@ -130,32 +113,22 @@ fn _escape<F: Fn(u8) -> bool>(raw: &[u8], escape_chars: F) -> Cow<[u8]> {
 /// Unescape a `&[u8]` and replaces all xml escaped characters ('&...;') into their corresponding
 /// value
 pub fn unescape(raw: &[u8]) -> Result<Cow<[u8]>, EscapeError> {
-    do_unescape(raw, None)
+    unescape_with(raw, |_| None)
 }
 
 /// Unescape a `&[u8]` and replaces all xml escaped characters ('&...;') into their corresponding
-/// value, using a dictionnary of custom entities.
+/// value, using a dictionary of custom entities.
 ///
 /// # Pre-condition
 ///
-/// The keys and values of `custom_entities`, if any, must be valid UTF-8.
-pub fn unescape_with<'a>(
+/// The implementation of `lookup_custom_entity` is expected to operate over UTF-8 inputs.
+pub fn unescape_with<'a, 'b>(
     raw: &'a [u8],
-    custom_entities: &HashMap<Vec<u8>, Vec<u8>>,
-) -> Result<Cow<'a, [u8]>, EscapeError> {
-    do_unescape(raw, Some(custom_entities))
-}
-
-/// Unescape a `&[u8]` and replaces all xml escaped characters ('&...;') into their corresponding
-/// value, using an optional dictionary of custom entities.
-///
-/// # Pre-condition
-///
-/// The keys and values of `custom_entities`, if any, must be valid UTF-8.
-pub fn do_unescape<'a>(
-    raw: &'a [u8],
-    custom_entities: Option<&HashMap<Vec<u8>, Vec<u8>>>,
-) -> Result<Cow<'a, [u8]>, EscapeError> {
+    lookup_custom_entity: impl Fn(&'b [u8]) -> Option<&'b str>,
```

Comment:
You need to use [Higher-Rank Trait Bounds](https://doc.rust-lang.org/nomicon/hrtb.html#higher-rank-trait-bounds-hrtbs) to fix lifetime issues.

Also it is better to name parameter `resolve_entity`

Reply:
Seems like there's still a lot of compile errors.  I cannot tell, if it is trying to get me to annotate this signature on top of the closure as well, but if so that would be a very user-unfriendly API.  Or maybe I'm just missing something (relatively) simple to fix.

Reply:
It looks like some of the errors are actually Rust issues.. https://github.com/rust-lang/rust/issues/75791

Reply:
Not all of them though.  Seems like Rust also has a difficult time recognizing that in each individual iteration of the loop only a single mutable or non-mutable reference can exist.

Reply:
Corrent implementation requires a bit more time than I expected, so I will postpone the task for a while. This requires a big work for review of encoding support, so not before 0.25.0